### PR TITLE
Fix torch.compile graph break in torch backend's dot_product_attention

### PR DESCRIPTION
**Pull request by pctablet505**
## Summary

`torch.compile` cannot trace the flash attention probe in the torch backend's
`dot_product_attention`, so it breaks the graph at every attention call. On a
six-layer transformer forward pass this produces 19 graph breaks and 20 separate
graphs.

`MultiHeadAttention` routes through `ops.dot_product_attention` when dropout is
0, attention scores are not returned, and inputs are 4D — the default path for
most models.

## Root cause

`dot_product_attention` calls `_can_use_flash_attention` when `flash_attention`
is `None` (the default), and again with `raise_error=True` when it is explicitly
`True`. That helper constructs a pybind11 `SDPAParams` object to probe kernel
support:

```python
spda_params = SDPAParams(query, key, value, mask, 0.0, is_causal, False)
```

Dynamo cannot trace a pybind11 constructor, so tracing stops there and resumes
afterwards — once per attention call.

## Change

Report flash attention as unavailable while tracing, so the probe is never
reached and the call takes its ordinary non-flash branch. `scaled_dot_product_attention`
then selects the kernel through its own dispatcher, which is what it does on the
non-flash branch already. `dot_product_attention` itself is unchanged — the whole
change is a 12-line early return at the top of `_can_use_flash_attention`.

**This applies to auto-detection only.** An explicit `flash_attention=True` still
runs the probe, still breaks the graph, and still raises
`Flash attention is not supported with the provided inputs` — deliberately. Making
the guard raise instead does not work: dynamo discards an exception raised while
tracing and re-runs the frame eagerly, where `is_compiling()` is `False`, so the
message never reaches the user and the frame is de-optimized rather than split.

Eager is unchanged by construction: `torch.compiler.is_compiling()` is `False`
outside `torch.compile`, so the probe and the pinned flash-attention kernel
context still run exactly as before.

## Results

`torch._dynamo.explain`, float32 unless noted. These counts are deterministic,
and are identical on an RTX 4050 (sm_89, torch 2.13.0+cu130) and an RTX PRO 6000
Blackwell (sm_120, torch 2.13.0+cu132):

| `flash_attention` | | graphs | breaks | ops captured |
|---|---|---|---|---|
| `None` (default) | before | 20 | 19 | 178 |
| `None` (default) | after | **1** | **0** | **367** |
| `True` | before | 2 | 1 | 16 |
| `True` | after | 2 | 1 | 16 |

Op count rises because a fragmented graph only captures a fraction of the model;
with one graph, dynamo captures all of it.

The `None` rows are the six-layer model below. The `True` rows are a single
`ops.dot_product_attention` call in float16, since float32 with
`flash_attention=True` raises on both master and this branch.

<details><summary>harness</summary>

```python
import keras, torch
from keras import layers
B, S, DM, H, LAYERS = 8, 128, 512, 8, 6
inp = keras.Input(shape=(S, DM)); x = inp
for _ in range(LAYERS):
    a = layers.MultiHeadAttention(num_heads=H, key_dim=DM // H)(x, x)
    x = layers.LayerNormalization()(x + a)
    f = layers.Dense(DM * 2, activation="relu")(x)
    x = layers.LayerNormalization()(x + layers.Dense(DM)(f))
model = keras.Model(inp, x).to("cuda")
t = torch.randn(B, S, DM, device="cuda")
print(torch._dynamo.explain(lambda z: model(z, training=False))(t))
```
</details>

Wall clock for the same six-layer forward, float32, on the RTX PRO 6000
Blackwell. Median of 8 A/B pairs with the order alternated per pair, each pair
30 repeats of 20 iterations timed with CUDA events. Running the base against an
identical checkout of itself measures run-to-run noise at 1.7% eager and 7.6%
compiled, putting the significance threshold at 2.5% and 11.4% respectively:

| | eager | compiled |
|---|---|---|
| before | 5.98 ms | 8.05 ms |
| after | 5.98 ms | **0.98 ms** |

Compiled is 8.2x faster, all 8 pairs agreeing in direction, 1.03x spread across
them. Eager lands at 1.005x, inside its threshold. Before this change,
`torch.compile` makes this model slower than its own eager forward.

## Behavior

Kernel dispatch is unchanged. Measured across fp16 / bf16 / fp32, with and without
a mask, and at head_dim 64, 128 and 320, the kernel selected by
`scaled_dot_product_attention` is identical to master in every case, in both eager
and compiled mode. On sm_120 that spans 72 configurations, with no difference in
any of them.

Under compile the auto-detect path no longer pins the kernel via
`sdpa_kernel([FLASH_ATTENTION])`, leaving the selection to the dispatcher. Where
the dispatcher lands on the kernel the pin would have chosen, output is
bit-identical: on sm_120, `max|compiled − eager|` is exactly 0.0 across 18
configurations spanning fp32 / fp16 / bf16, head_dim 64 and 128, plain, masked
and causal. Where it lands elsewhere the two can differ slightly; on sm_89 fp16
shows about 1.2e-04 and fp32 about 6e-08.

## Tests

`keras/src/backend/torch/nn_test.py` adds one parameterized test that compiles
`ops.dot_product_attention` with `fullgraph=True`, so a graph break becomes an
error, and compares against eager across six argument shapes: plain, causal,
masked, masked-and-causal, bias, and custom scale.

All six fail against master and pass with this change, both on a CPU-only torch
build as CI runs it and on CUDA (sm_120).

Scope, stated honestly: the cases are float32 and `flash_attention=None`, so they
cover the auto-detect path only. Grouped-query inputs, explicit `flash_attention`
values, and the hardware-dependent flash-vs-dispatcher divergence above are not
covered, and CI runs CPU-only so the kernel-dispatch matrix is not covered there
either.

- [x]  I am a human, and not a bot.
- [x]  I will be responsible for responding to review comments in a timely manner.
- [x]  I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: keras/src/backend/torch/nn.py [hunk 1]
```diff
@@ -1552,6 +1552,25 @@ def dot_product_attention(
         value = torch.repeat_interleave(value, repeats=groups, dim=1)
 
     if flash_attention is None:
+        if (
+            hasattr(torch.compiler, "is_compiling")
+            and torch.compiler.is_compiling()
+        ):
```
Comment (hertschuh) [inline keras/src/backend/torch/nn.py]: Instead of replicating the code to `scaled_dot_product_attention`, I would simply move this test at the beginning of `_can_use_flash_attention`.

This would be more true to the spirit of what the `flash_attention` argument means. If you compile, you cannot use flash attention, and the user should be aware of that.

File: keras/src/backend/torch/nn_test.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        )
+
+        for kwargs in (
+            {},
+            {"is_causal": True},
+            {"mask": mask},
+            {"mask": mask, "is_causal": True},
+            {"bias": bias},
+            {"scale": 0.125},
+        ):
```
Comment (hertschuh) [inline keras/src/backend/torch/nn_test.py]: Make this a parameterized test instead.

File: keras/src/backend/torch/nn_test.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+            def attend(kwargs=kwargs):
+                return ops.dot_product_attention(query, key, value, **kwargs)
```
Comment (hertschuh) [inline keras/src/backend/torch/nn_test.py]: Why is it called `attend`?

Also, I think you should pass `query`, `key`, `value` as arguments to `attend`. Otherwise, they can be inlined as constants during the compilation and you may not be exercising what you think you're exercising.

File: keras/src/backend/torch/nn_test.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+class DotProductAttentionCompileTest(testing.TestCase):
+    def _qkv(self, batch=2, seq=16, heads=4, head_dim=8):
+        rng = np.random.default_rng(0)
+        shape = (batch, seq, heads, head_dim)
+        return tuple(
+            ops.convert_to_tensor(rng.standard_normal(shape).astype("float32"))
+            for _ in range(3)
+        )
+
+    def test_compiles_without_graph_break(self):
```
Comment (hertschuh) [inline keras/src/backend/torch/nn_test.py]: Everything in this test is also done in the `test_compiled_matches_eager`. Just remove.

Also, why is it called "break"? To me break means it raises an error, but it doesn't.

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks for looking into this!

File: keras/src/backend/torch/nn_test.py (hunk 3 above)
Comment (pctablet505) [reply keras/src/backend/torch/nn_test.py]: The wrapper is gone entirely — it now compiles `ops.dot_product_attention` directly, so the inputs are genuine traced arguments and there is no wrapper left to name.

File: keras/src/backend/torch/nn_test.py (hunk 4 above)
Comment (pctablet505) [reply keras/src/backend/torch/nn_test.py]: Removed it. 

"Graph break" is PyTorch's term (`graph_break_count` in `torch._dynamo.explain`) and you're right that it doesn't raise on its own — dynamo just splits the graph and runs the untraceable part eagerly. Only `fullgraph=True` turns it into an error, so the name described the condition rather than the assertion. Bad name either way.

One caveat: test_compiled_matches_eager didn't have fullgraph=True, so it passed with the break present. I moved that flag onto the parameterized test, which now fails 6/6 on master and passes 6/6 with the fix.

File: keras/src/backend/torch/nn.py (hunk 1 above)
Comment (pctablet505) [reply keras/src/backend/torch/nn.py]: Moved it, `dot_product_attention` is now identical to master, and the change is just a guard at the top of `_can_use_flash_attention`.

Two corrections from measuring it on an RTX 4050 (sm_89), though:

"If you compile, you cannot use flash attention" isn't the case. With the probe skipped and fp16 inputs, `scaled_dot_product_attention` still dispatches to `pytorch_flash::flash_fwd_kernel (torch.profiler, probe executions 0)`. Only the pybind11 `SDPAParams` construction is untraceable, not the kernel, so returning False here costs no flash attention.

Making the user aware via `raise_error` doesn't work either; I tried it first. Dynamo discards an exception raised while tracing and re-runs the frame eagerly, where `is_compiling()` is `False`. So the probe runs anyway, the message never surfaces, and `op_count` drops to 1 from master's 16. Worse than the break it was removing.

So the guard only covers auto-detection (not `raise_error`). An explicit `flash_attention=True` still behaves exactly as on master and reports `"Flash attention is not supported with the provided inputs"`.

Comment (hertschuh) [reply keras/src/backend/torch/nn.py]: > "If you compile, you cannot use flash attention" isn't the case. With the probe skipped and fp16 inputs, `scaled_dot_product_attention` still dispatches to `pytorch_flash::flash_fwd_kernel (torch.profiler, probe executions 0)`. Only the pybind11 `SDPAParams` construction is untraceable, not the kernel, so returning False here costs no flash attention.

So what you're saying is that `scaled_dot_product_attention` is always and automatically "upgraded" to flash attention when you compile? So we cannot really apply `flash_attention=True/False`, it's always going to be True?

**Review by hertschuh — COMMENTED** [review]
See https://github.com/keras-team/keras/pull/23327#discussion_r3687800256

File: keras/src/backend/torch/nn.py (hunk 1 above)
Comment (pctablet505) [reply keras/src/backend/torch/nn.py]: Short answer: no. `flash_attention=True` still works exactly as before, and
`flash_attention=False` never turned flash attention off — that is already true
on master, in eager. This PR changes neither of it.

**What the flag does**

- `True` — pins the flash kernel and raises if the input cannot use it. This PR
  leaves it alone (the guard is `not raise_error`), so it still raises under
  `torch.compile`.
- `False` — does **not** forbid flash. It skips the pin and calls
  `scaled_dot_product_attention` with no `sdpa_kernel(...)` context, so PyTorch's
  dispatcher chooses — and for fp16/bf16 it chooses flash.
- `None` — probes, then does one of the above. Under compile this PR skips the
  probe, landing on exactly the same code path as `False`.

**Measured dispatch**, classified by the CUDA kernel that actually runs:

| input | `None` | `True` | `False` |
|---|---|---|---|
| fp16/bf16, no mask | flash | flash | **flash** |
| fp16/bf16, masked | mem-efficient | raises | mem-efficient |
| fp32 | mem-efficient | raises | mem-efficient |

Same in eager and compiled. I ran all 72 combinations — 3 dtypes × mask/no-mask ×
head_dim 64/128 × 3 flag values × eager/compiled — on master and on this branch:
**both pick the same kernel in every one, 0 differences.**

Two things follow from the bold cell:

- It is **master, in eager**: `flash_attention=False` runs
  `pytorch_flash::flash_fwd_kernel`. So "`False` does not disable flash" is
  pre-existing behavior, not something this PR introduces.
- fp32 never uses flash at all (flash requires fp16/bf16), so on the default
  float32 path this change alters the selected kernel in zero cases.

If `False` is meant to genuinely forbid flash, that is a real gap — but a
separate one on master, and I am happy to open a PR for it.

**Wall clock.** The PR body omits this because my previous machine could not hold
a stable measurement. This one can. Six-layer transformer forward, float32,
median of 8 alternating A/B pairs against the merge-base:

| | eager | compiled |
|---|---|---|
| without this change | 5.98 ms | 8.05 ms |
| with this change | 5.98 ms | **0.98 ms** |

- compiled: 8.2x faster, 8/8 pairs agreeing, per-pair spread 1.03x
- eager: unchanged (1.005x, inside a 2.5% noise floor); the guard costs 0.1 µs
  per call
- without the change, `torch.compile` makes this model *slower* than eager

**Review by hertschuh — APPROVED** [review]
Thanks for the fix!
